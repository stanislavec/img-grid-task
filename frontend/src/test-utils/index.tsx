import React, { ReactElement } from 'react';
import { createBrowserRouter, createMemoryRouter, RouterProvider } from 'react-router-dom';

export const withRouter = (Component: ReactElement<unknown, any>, path: string = '/') => {
  const router = createBrowserRouter([
    {
      path,
      element: Component,
    },
  ]);

  return <RouterProvider router={router} />;
};

export const withMemotyRouter = (
  Component: ReactElement<unknown, any>,
  { route = '/', path = '/' } = {},
) => {
  const router = createMemoryRouter(
    [
      {
        path,
        element: Component,
      },
    ],
    {
      initialEntries: [route],
    },
  );
  return <RouterProvider router={router} />;
};
