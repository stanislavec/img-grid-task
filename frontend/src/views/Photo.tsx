import React, { useMemo } from 'react';
import { useLocation, useParams } from 'react-router-dom';
import Button from 'components/Button';
import Loader from 'components/Loader';
import usePhotoContext from 'context/useImageContext';
import scss from './Photo.module.scss';

function Photo() {
  const { photos, loading, favorites, removeFromFavorite } = usePhotoContext();

  const { id } = useParams();

  const { state } = useLocation();

  const photo = state?.photo || photos?.find((ph) => ph.id === id);

  const isFavorite = useMemo(
    () => favorites.some((favorite) => favorite.id === id),
    [favorites, id],
  );

  if (loading) return <Loader />;

  if (!photo) return <>Seems, image does not exist</>;

  return (
    <div className={scss.photo}>
      <img src={photo.download_url} alt={photo.url}></img>
      {isFavorite ? (
        <Button onClick={() => removeFromFavorite(photo.id)}>Remove from favorites</Button>
      ) : (
        <>This photo is not the favorite one</>
      )}
    </div>
  );
}

export default Photo;
