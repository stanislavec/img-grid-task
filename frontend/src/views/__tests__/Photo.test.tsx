import { fireEvent, render, screen } from '@testing-library/react';
import usePhotoContext from 'context/useImageContext';
import { withMemotyRouter } from 'test-utils';
import Photo from '../Photo';
import PHOTOS from './__mocks__/photo';

jest.mock('context/useImageContext');

const mockUseNavigate = jest.fn();

const mockUsePhotoContext = usePhotoContext as jest.MockedFunction<typeof usePhotoContext>;

describe('Photo', () => {
  beforeEach(() => jest.clearAllMocks());

  it('is defined', () => {
    mockUsePhotoContext.mockReturnValue({
      photos: [],
      fetchNext: () => null,
      loading: false,
      addToFavorite: () => null,
      removeFromFavorite: () => null,
      favorites: [],
    });
    const elem = render(withMemotyRouter(<Photo />, { route: '/photo/5', path: '/photo/:id' }));
    expect(elem.baseElement).toBeDefined();
  });

  it('matches snapshot if photo exists', () => {
    mockUsePhotoContext.mockReturnValue({
      photos: PHOTOS,
      fetchNext: () => null,
      loading: false,
      addToFavorite: () => null,
      removeFromFavorite: () => null,
      favorites: PHOTOS.slice(0, 10),
    });
    const elem = render(withMemotyRouter(<Photo />, { route: '/photo/5', path: '/photo/:id' }));
    expect(elem.container).toMatchSnapshot();
  });

  it('matches snapshot if photo does not exists', () => {
    mockUsePhotoContext.mockReturnValue({
      photos: [],
      fetchNext: () => null,
      loading: false,
      addToFavorite: () => null,
      removeFromFavorite: () => null,
      favorites: [],
    });
    const elem = render(withMemotyRouter(<Photo />, { route: '/photo/5', path: '/photo/:id' }));
    expect(elem.container).toMatchSnapshot();
  });

  it('matches snapshot if photo exists but not favorite', () => {
    mockUsePhotoContext.mockReturnValue({
      photos: PHOTOS,
      fetchNext: () => null,
      loading: false,
      addToFavorite: () => null,
      removeFromFavorite: () => null,
      favorites: [],
    });
    const elem = render(withMemotyRouter(<Photo />, { route: '/photo/5', path: '/photo/:id' }));
    expect(elem.container).toMatchSnapshot();
  });

  it('handles remove from favorites button click', () => {
    const removeFromFavorite = jest.fn();
    mockUsePhotoContext.mockReturnValue({
      photos: PHOTOS,
      fetchNext: () => null,
      loading: false,
      addToFavorite: () => null,
      removeFromFavorite,
      favorites: PHOTOS.slice(0, 10),
    });

    render(withMemotyRouter(<Photo />, { route: '/photo/5', path: '/photo/:id' }));

    const button = screen.queryByText('Remove from favorites');

    if (!button) throw Error();

    fireEvent.click(button);

    expect(removeFromFavorite).toBeCalledTimes(1);
  });
});
