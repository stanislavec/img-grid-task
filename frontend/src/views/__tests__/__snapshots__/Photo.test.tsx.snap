// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Photo matches snapshot if photo does not exists 1`] = `
<div>
  Seems, image does not exist
</div>
`;

exports[`Photo matches snapshot if photo exists 1`] = `
<div>
  <div
    class="photo"
  >
    <img
      alt="https://unsplash.com/photos/LF8gK8-HGSg"
      src="https://picsum.photos/id/5/5000/3334"
    />
    <button
      class="button"
    >
      Remove from favorites
    </button>
  </div>
</div>
`;

exports[`Photo matches snapshot if photo exists but not favorite 1`] = `
<div>
  <div
    class="photo"
  >
    <img
      alt="https://unsplash.com/photos/LF8gK8-HGSg"
      src="https://picsum.photos/id/5/5000/3334"
    />
    This photo is not the favorite one
  </div>
</div>
`;
